### PR TITLE
AMDGPU/GlobalISel: Fix inst-selection of ballot

### PR DESCRIPTION
**Pull request by petar-avramovic**
Both input and output of ballot are lane-masks:
result is lane-mask with 'S32/S64 LLT and SGPR bank'
input is lane-mask with 'S1 LLT and VCC reg bank'.
Ballot copies bits from input lane-mask for
all active lanes and puts 0 for inactive lanes.
GlobalISel did not set 0 in result for inactive lanes
for non-constant input.

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp [hunk 1]
```diff
@@ -1429,34 +1429,57 @@ bool AMDGPUInstructionSelector::selectBallot(MachineInstr &I) const {
...
-          .addReg(SrcReg);
-      return;
-    }
+  const auto BuildAnd = [&](unsigned Opcode, Register Dst, Register Src,
+                            Register Exec) {
+    auto And = BuildMI(*BB, &I, DL, TII.get(Opcode), Dst)
+                   .addReg(Src)
+                   .addReg(Exec)
+                   .setOperandDead(3); // Dead scc
+    constrainSelectedInstRegOperands(*And, TII, TRI, RBI);
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp:1438]: Should check return value

File: llvm/test/CodeGen/AMDGPU/GlobalISel/llvm.amdgcn.ballot.i32.ll [hunk 2]
```diff
@@ -65,7 +68,8 @@ define amdgpu_cs i32 @compare_int_with_constant(i32 %x) {
 define amdgpu_cs i32 @compare_floats(float %x, float %y) {
 ; CHECK-LABEL: compare_floats:
 ; CHECK:       ; %bb.0:
-; CHECK-NEXT:    v_cmp_gt_f32_e64 s0, v0, v1
+; CHECK-NEXT:    v_cmp_gt_f32_e32 vcc_lo, v0, v1
+; CHECK-NEXT:    s_and_b32 s0, vcc_lo, exec_lo
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/llvm.amdgcn.ballot.i32.ll:72]: v_cmp already treats inactive lanes as 0. SelectionDAG does not introduce the extra instruction

File: llvm/test/CodeGen/AMDGPU/llvm.amdgcn.ballot.i32.ll [hunk 3]
```diff
@@ -522,3 +522,76 @@ true:
...
+; GFX10-NEXT:  ; %bb.2: ; %Flow
+; GFX10-NEXT:    s_andn2_saveexec_b32 s1, s1
+; GFX10-NEXT:  ; %bb.3: ; %A
+; GFX10-NEXT:    v_cmp_ne_u32_e32 vcc_lo, 0, v2
+; GFX10-NEXT:    s_andn2_b32 s0, s0, exec_lo
+; GFX10-NEXT:    s_and_b32 s2, vcc_lo, exec_lo
+; GFX10-NEXT:    s_or_b32 s0, s0, s2
+; GFX10-NEXT:  ; %bb.4: ; %exit
+; GFX10-NEXT:    s_or_b32 exec_lo, exec_lo, s1
+; GFX10-NEXT:    v_cndmask_b32_e64 v2, 0, 1, s0
```
Comment (petar-avramovic) [inline llvm/test/CodeGen/AMDGPU/llvm.amdgcn.ballot.i32.ll:548]: for context this is what sdag does.

non-constant ballot
tsrc:    i1 = ...
tballot: i32 = llvm.amdgcn.ballot tsrc

is selected as (by default, when combine does not kick in)
tzext: i32 = zero_extend tsrc
tballot: i32 = SETCC tzext, Constant:i32<0>, setne:ch

Note the uppercase SETCC and i32 type, regular compare is "i1 = setcc"

most common case is with regular compare input that gets folded(*)
tsrc:    i1 = setcc t_op0, t_op1
tballot: i32 = llvm.amdgcn.ballot tsrc

tballot: i32 = SETCC t_op0, t_op1

Case that is missing on GlobalISel is when inactive lanes need to be zeroed by ballot.
SDag does this by selecting tzext: i32 = zero_extend tsrc into select
and then comparing it with zero - compare result in vcc is set to 0 for inactive lanes.

(*) In theory this might be wrong in case folded compare was from another block with different exec mask but might not be possible since sdag works block by block?

GlobalISel could avoid adding 'AND with exec' when input is compare result that was calculated with same exec mask (compared to current exec - used by ballot)

File: llvm/test/CodeGen/AMDGPU/GlobalISel/llvm.amdgcn.ballot.i32.ll (hunk 2 above)
Comment (petar-avramovic) [reply llvm/test/CodeGen/AMDGPU/GlobalISel/llvm.amdgcn.ballot.i32.ll:72]: Code is worse at the moment but correct. I don't see simple way to fix it.
Maybe if lower ballot earlier but most probably need new regbankselect first then remove extra AND in regbankcombiner.

**Comment by petar-avramovic** [comment]
Maybe it is possible to not make AND in some cases. Will look into it.

**Comment by petar-avramovic** [comment]
Ping. Fixed most of the cases where compare was no folded, there is a comment that justifies why I think it is correct to fold.

File: llvm/test/CodeGen/AMDGPU/GlobalISel/llvm.amdgcn.ballot.i32.ll [hunk 4]
```diff
@@ -89,7 +90,8 @@ define amdgpu_cs i32 @branch_divergent_ballot_ne_zero_non_compare(i32 %v) {
 ; CHECK:       ; %bb.0:
 ; CHECK-NEXT:    v_and_b32_e32 v0, 1, v0
 ; CHECK-NEXT:    v_cmp_ne_u32_e32 vcc_lo, 0, v0
-; CHECK-NEXT:    s_cmp_eq_u32 vcc_lo, 0
+; CHECK-NEXT:    s_and_b32 s0, vcc_lo, exec_lo
```
[the remark this answers is not in the document]
Comment (petar-avramovic) [reply llvm/test/CodeGen/AMDGPU/GlobalISel/llvm.amdgcn.ballot.i32.ll:93]: This is really not obvious, Input is "vgpr to vcc trunc" that is selected as compare. Will become better after new reg bank select.

File: llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp [hunk 5]
```diff
@@ -1429,34 +1429,129 @@ bool AMDGPUInstructionSelector::selectBallot(MachineInstr &I) const {
...
+      SrcMI = MRI->getVRegDef(SrcMI->getOperand(1).getReg());
+
+    if (SrcMI->getOpcode() == AMDGPU::G_ICMP ||
+        SrcMI->getOpcode() == AMDGPU::G_FCMP)
+      return SrcMI;
+    return nullptr;
+  };
+
+  const auto FoldCmp = [&](Register Dst, MachineInstr *CmpMI) -> bool {
+    // Fold ballot of a compare. Active lanes when the ballot is executed need
```
[the remark this answers is not in the document]
Comment (petar-avramovic) [reply llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp:1445]: We can do better. There is no need to check anything if compare is used directly. 

For example:
```
%bb.entry
%cmp = ...
s_and_saveexec_b32 ...

%bb.divergent_block
 %res = ballot(%cmp)
```
Works because active lanes in `%bb.divergent_block` are subset of active lanes in `%bb.entry`, so "sinking compare" to other block works according to ballot description (put 0 in inactive lanes).

the only potential problem is when 

```
%bb.entry
s_and_saveexec_b32 ...

%bb.divergent_block
 %cmp = ...

; %bb.merge.control.flow
 s_or_b32 %exec.from.bb.divergent_block, ...
 %res = ballot(%cmp)
```
Lane could be inactive in `%bb.divergent_block`
but active in `%bb.merge.control.flow,` and ballot result could be 1 instead of 0.
However that case is not written correctly. ballot can't use such compare directly, it has to use phi.

```
%bb.entry
s_and_saveexec_b32 ...

%bb.divergent_block
 **%cmp** = ...
...
; %bb.merge.control.flow
 s_or_b32 %exec.from.bb.divergent_block, ...
 %phi = phi i1 [ **%cmp**, %bb.divergent_block ], [ ..., ...]
 %res = ballot(%phi)
```

So in summary we do same thing as sdag if input is compare (sink it). Otherwise select and with exec.
This is better then what selection dag does when input is not compare (sdag does select 0, 1 (lane mask to vgpr) + compare vgpr with 0)

File: llvm/test/CodeGen/AMDGPU/GlobalISel/llvm.amdgcn.struct.ptr.buffer.store.format.f32.ll [hunk 6]
```diff
@@ -132,7 +132,8 @@ define amdgpu_ps void @struct_ptr_buffer_store_format_f32__sgpr_val__vgpr_rsrc__
   ; CHECK-NEXT:   [[V_READFIRSTLANE_B32_4:%[0-9]+]]:sreg_32 = V_READFIRSTLANE_B32 [[COPY7]], implicit $exec
   ; CHECK-NEXT:   [[V_CMP_EQ_U32_e64_:%[0-9]+]]:sreg_64_xexec = V_CMP_EQ_U32_e64 [[V_READFIRSTLANE_B32_4]], [[COPY7]], implicit $exec
   ; CHECK-NEXT:   [[S_AND_B64_1:%[0-9]+]]:sreg_64_xexec = S_AND_B64 [[S_AND_B64_]], [[V_CMP_EQ_U32_e64_]], implicit-def dead $scc
-  ; CHECK-NEXT:   [[S_AND_SAVEEXEC_B64_:%[0-9]+]]:sreg_64_xexec = S_AND_SAVEEXEC_B64 killed [[S_AND_B64_1]], implicit-def $exec, implicit-def $scc, implicit $exec
+  ; CHECK-NEXT:   [[S_AND_B64_2:%[0-9]+]]:sreg_64_xexec = S_AND_B64 [[S_AND_B64_1]], $exec, implicit-def dead $scc
+  ; CHECK-NEXT:   [[S_AND_SAVEEXEC_B64_:%[0-9]+]]:sreg_64_xexec = S_AND_SAVEEXEC_B64 killed [[S_AND_B64_2]], implicit-def $exec, implicit-def $scc, implicit $exec
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/llvm.amdgcn.struct.ptr.buffer.store.format.f32.ll:136]: Nearly all of the test changes are regressing the waterfall loop case, for the ballot from regbankselect. Can we do anything about that? Intuitively we should be able to use the raw value

File: llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp [hunk 7]
```diff
@@ -1429,34 +1432,131 @@ bool AMDGPUInstructionSelector::selectBallot(MachineInstr &I) const {
...
+    };
+
+    MachineInstrBuilder Cmp;
+    if (CmpMI->getOpcode() == AMDGPU::G_ICMP) {
+      Cmp = BuildMI(*BB, &I, DL, TII.get(CmpOpcode), Dst);
+      Cmp.addReg(constrainToVGPR(Src0Reg, Cmp))
+          .addReg(constrainToVGPR(Src1Reg, Cmp));
+    } else {
+      auto [Src0, Src0Mods] = selectVOP3ModsImpl(Src0Reg);
+      auto [Src1, Src1Mods] = selectVOP3ModsImpl(Src1Reg);
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp:1482]: Need tests with source modifiers?

File: llvm/test/CodeGen/AMDGPU/GlobalISel/llvm.amdgcn.struct.ptr.buffer.store.format.f32.ll (hunk 6 above)
Comment (petar-avramovic) [reply llvm/test/CodeGen/AMDGPU/GlobalISel/llvm.amdgcn.struct.ptr.buffer.store.format.f32.ll:136]: Will rewrite the patch next week, need to look through AND and OR and see if some of the values was made with same exec(going back to your suggestion to just check for same block). I was thinking to simply do 'copy' instead of sinking compare or 'AND with exec'

File: llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp [hunk 8]
```diff
@@ -1429,34 +1432,131 @@ bool AMDGPUInstructionSelector::selectBallot(MachineInstr &I) const {
...
-  const auto BuildCopy = [&](Register SrcReg) {
-    if (Size == STI.getWavefrontSize()) {
-      BuildMI(*BB, &I, DL, TII.get(AMDGPU::COPY), DstReg)
-          .addReg(SrcReg);
-      return;
+  const auto getCmpInput = [&]() -> MachineInstr * {
+    MachineInstr *SrcMI = getDefIgnoringCopies(I.getOperand(2).getReg(), *MRI);
+    // Try to fold sgpr compare.
+    if (SrcMI->getOpcode() == AMDGPU::G_TRUNC)
+      SrcMI = MRI->getVRegDef(SrcMI->getOperand(1).getReg());
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp:1439]: This pattern probably shouldn't reach selection

File: llvm/test/CodeGen/AMDGPU/GlobalISel/llvm.amdgcn.ballot.i32.ll [hunk 9]
```diff
@@ -419,3 +424,80 @@ true:
...
+; GFX11-NEXT:  ; %bb.3: ; %A
+; GFX11-NEXT:    v_cmp_le_u32_e32 vcc_lo, 1, v2
+; GFX11-NEXT:    s_and_not1_b32 s0, s0, exec_lo
+; GFX11-NEXT:    s_and_b32 s2, exec_lo, vcc_lo
+; GFX11-NEXT:    s_or_b32 s0, s0, s2
+; GFX11-NEXT:  ; %bb.4: ; %exit
+; GFX11-NEXT:    s_or_b32 exec_lo, exec_lo, s1
+; GFX11-NEXT:    s_and_b32 s0, s0, exec_lo
+; GFX11-NEXT:    v_mov_b32_e32 v2, s0
+; GFX11-NEXT:    global_store_b32 v[0:1], v2, off
```
Comment (petar-avramovic) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/llvm.amdgcn.ballot.i32.ll:482]: This is ballot source that is not compare (source is phi), compared to what sdag does:
https://github.com/llvm/llvm-project/pull/109986/commits/0c0e21bf407bb4616e7283befec8ac0aec361ee3#diff-8048303f4c9f4c844e010c047a71b7bfcfb6c612f0faf5d9a89104acb11ee39fR571-R576

File: llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp [hunk 10]
```diff
@@ -1413,50 +1413,97 @@ bool AMDGPUInstructionSelector::selectIntrinsicCmp(MachineInstr &I) const {
...
+                  MachineBasicBlock *MBB) {
+  MachineInstr *MI = MRI.getVRegDef(Reg);
+  // Look through copies, truncs and anyext. TODO: just copies
+  while (MI->getOpcode() == AMDGPU::COPY ||
+         MI->getOpcode() == AMDGPU::G_TRUNC ||
+         MI->getOpcode() == AMDGPU::G_ANYEXT) {
+    Reg = MI->getOperand(1).getReg();
+    if (!Reg.isVirtual())
+      return false;
+    MI = MRI.getVRegDef(Reg);
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp:1429]: Maybe should break if the block changed

File: llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp [hunk 11]
```diff
@@ -1413,50 +1413,97 @@ bool AMDGPUInstructionSelector::selectIntrinsicCmp(MachineInstr &I) const {
   return true;
 }
 
+// Ballot has to zero bits in input lane-mask that are zero in current exec,
+// Done as AND with exec. For inputs that are results of instruction that
+// implicitly use same exec, for example compares in same basic block, use copy.
+bool isBallotCopy(Register Reg, MachineRegisterInfo &MRI,
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp:1419]: static

File: llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp [hunk 12]
```diff
@@ -1413,50 +1413,97 @@ bool AMDGPUInstructionSelector::selectIntrinsicCmp(MachineInstr &I) const {
...
+                  MachineBasicBlock *MBB) {
+  MachineInstr *MI = MRI.getVRegDef(Reg);
+  // Look through copies, truncs and anyext. TODO: just copies
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp:1422]: What situations are there redundant copies that need to be looked through? I would assume any trivial copies to be optimized out

File: llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp (hunk 11 above)
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp:1419]: Name is misleading, this is looking for a vector compare

File: llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp (hunk 12 above)
Comment (petar-avramovic) [reply llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp:1422]: non-trivial cross bank copies scc to vcc for example.

Comment (arsenm) [reply llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp:1422]: That would be at most one copy? 

Also thinking about whether such a trivially uniform ballot should have folded out already

**Review by arsenm — APPROVED** [review]

File: llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp [hunk 13]
```diff
@@ -1413,50 +1413,101 @@ bool AMDGPUInstructionSelector::selectIntrinsicCmp(MachineInstr &I) const {
...
+  if (MI->getOpcode() == AMDGPU::COPY) {
+    auto DstRB = MRI.getRegBankOrNull(MI->getOperand(0).getReg());
+    auto SrcRB = MRI.getRegBankOrNull(MI->getOperand(1).getReg());
+    if (DstRB && SrcRB && DstRB->getID() == AMDGPU::VCCRegBankID &&
+        SrcRB->getID() == AMDGPU::SGPRRegBankID)
+      return true;
+  }
+
+  // Lane mask generated using compare with same exec.
+  if (isa<GAnyCmp>(MI))
```
Comment (nhaehnle) [inline llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp:1436]: This is now missing a check that MI is in MBB.

File: llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp [hunk 14]
```diff
@@ -1413,50 +1413,101 @@ bool AMDGPUInstructionSelector::selectIntrinsicCmp(MachineInstr &I) const {
...
 
+// Ballot has to zero bits in input lane-mask that are zero in current exec,
+// Done as AND with exec. For inputs that are results of instruction that
+// implicitly use same exec, for example compares in same basic block or SCC to
+// VCC copy, use copy.
+static bool isLaneMaskFromSameBlock(Register Reg, MachineRegisterInfo &MRI,
+                                    MachineBasicBlock *MBB) {
+  MachineInstr *MI = MRI.getVRegDef(Reg);
+  if (MI->getParent() != MBB)
+    return false;
```
Comment (petar-avramovic) [inline llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp:1424]: same MBB check is here. Deleted the loop
